pkg/istio-agent/xds_proxy.go

```diff
@@ -363,6 +374,44 @@ func (p *XdsProxy) handleUpstreamResponse(con *ProxyConnection) {
 					TypeUrl:       v3.NameTableType,
 					ResponseNonce: resp.Nonce,
 				}
+			case resource.ListenerType:
+				// intercept listeners, replace Wasm remote load with local file path.
+				// TODO: deal with fail open, or not?
+				// TODO: add parallel download
+				proxyLog.Info("intercept listener")
+				sendNack := false
+				for i, r := range resp.Resources {
+					var l listener.Listener
+					if err := ptypes.UnmarshalAny(r, &l); err != nil {
+						log.Errorf("failed to unmarshall listener: %v", err)
+						continue
+					}
+					if found, success := wasmconvert.ConvertWasmFilter(&l, p.wasmCache); !found {
+						log.Debugf("no need to marshal listener again")
+						continue
+					} else if !success {
+						// Send Nack
+						sendNack = true
+						break
+					}
+					nl, err := ptypes.MarshalAny(&l)
+					if err != nil {
+						log.Errorf("failed to marshall new listener: %v", err)
+					}
+					resp.Resources[i] = nl
+				}
+				if sendNack {
+					con.requestsChan <- &discovery.DiscoveryRequest{
+						VersionInfo:   p.lastLDSAckVersion,
+						TypeUrl:       resource.ListenerType,
+						ResponseNonce: resp.Nonce,
+						ErrorDetail: &grpcStatus.Status{
+							Message: "fail to fetch wasm module",
```

Comment:
More details on what failed?add  to metrics

Reply:
Yeah for sure. This change needs many polishes if we decide to chase this down further.